src/segmentation.rs

```diff
@@ -89,41 +279,111 @@ pub fn select_segment<T: Pixel>(
+    let x_in_imp_b = ((tile_offset.x + plane_offset.x) << plane_cfg.xdec)
+      as usize
+      / IMPORTANCE_BLOCK_SIZE;
+    let y_in_imp_b = ((tile_offset.y + plane_offset.y) << plane_cfg.ydec)
+      as usize
+      / IMPORTANCE_BLOCK_SIZE;
+    let w_in_imp_b = &fi.coded_frame_data.as_ref().unwrap().w_in_imp_b;
+
+    let seg =
+      ts.segmentation.segmentation_mask[y_in_imp_b * w_in_imp_b + x_in_imp_b];
+    return seg..=seg;
+  }
+
+  // With temporal RDO only (no AQ), compute the index now.
   let frame_bo = ts.to_frame_block_offset(tile_bo);
+  let sidx =
+    get_temporal_rdo_sidx(fi, ts, frame_bo, bsize, segment_2_is_lossless)
+      as u8;
+
+  sidx..=sidx
+}
+
+fn get_temporal_rdo_sidx<T: Pixel>(
+  fi: &FrameInvariants<T>, ts: &TileStateMut<'_, T>,
+  frame_bo: PlaneBlockOffset, bsize: BlockSize, segment_2_is_lossless: bool,
+) -> usize {
   let scale = spatiotemporal_scale(fi, frame_bo, bsize);
 
   // TODO: Replace this calculation with precomputed scale thresholds.
-  let seg_ac_q: ArrayVec<_, 3> = if fi.enable_segmentation {
-    use crate::quantize::ac_q;
-    (0..=2)
-      .map(|sidx| {
-        ac_q(
-          (fi.base_q_idx as i16
-            + ts.segmentation.data[sidx][SegLvl::SEG_LVL_ALT_Q as usize])
-            .max(0)
-            .min(255) as u8,
-          0,
-          fi.sequence.bit_depth,
-        )
-      })
-      .collect()
-  } else {
-    Default::default()
-  };
+  let seg_ac_q: ArrayVec<_, 3> = seg_ac_q(fi, ts);
 
-  let sidx = if scale.mul_u64(seg_ac_q[1] as u64) < seg_ac_q[0] as u64 {
+  if scale.mul_u64(seg_ac_q[1] as u64) < seg_ac_q[0] as u64 {
+    // Use higher qindex
     1
   } else if !segment_2_is_lossless
     && scale.mul_u64(seg_ac_q[2] as u64) > seg_ac_q[0] as u64
   {
+    // Use lower qindex
     2
   } else {
+    // Do not modify qindex
     0
-  };
+  }
+}
 
-  sidx..=sidx
+fn seg_ac_q<T: Pixel>(
+  fi: &FrameInvariants<T>, ts: &TileStateMut<T>,
+) -> ArrayVec<i16, 3> {
+  (0..=2)
+    .map(|sidx| {
+      ac_q(
+        (fi.base_q_idx as i16
+          + ts.segmentation.data[sidx][SegLvl::SEG_LVL_ALT_Q as usize])
+          .max(0)
+          .min(255) as u8,
+        0,
+        fi.sequence.bit_depth,
+      )
+    })
+    .collect()
+}
+
+fn apply_temporal_rdo_adjustments<T: Pixel>(
+  segments: &mut [u8], fi: &FrameInvariants<T>, fs: &mut FrameState<T>,
+) {
+  let w_in_imp_b = fi.coded_frame_data.as_ref().unwrap().w_in_imp_b;
+  let ts = fs.as_tile_state_mut();
+  for (i, segment) in segments.iter_mut().enumerate() {
+    // The activity masks are in 16x16 blocks, so we will iterate by 16x16 here as well.
+    let frame_bo = PlaneBlockOffset(BlockOffset {
+      x: (i % w_in_imp_b) << 1,
+      y: (i / w_in_imp_b) << 1,
+    });
+
+    // Apply temporal RDO bias
+    let trdo_sidx =
+      get_temporal_rdo_sidx(fi, &ts, frame_bo, BlockSize::BLOCK_16X16, false);
+    // TODO: Use more gradual adjustments
+    match trdo_sidx {
```

Comment:
In particular this seems not equivalent to what we were doing before. Before for temporal RDO, we had 3 segments, +1, -1 and 0. Now, we are bumping segment numbers up and down amounts in the amount of aq_strength for temporal RDO, which seems wrong.